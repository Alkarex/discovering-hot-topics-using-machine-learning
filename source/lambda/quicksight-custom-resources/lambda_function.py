#!/usr/bin/env python
######################################################################################################################
#  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           #
#                                                                                                                    #
#  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    #
#  with the License. A copy of the License is located at                                                             #
#                                                                                                                    #
#      http://www.apache.org/licenses/LICNSE-2.0                                                                     #
#                                                                                                                    #
#  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES #
#  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    #
#  and limitations under the License.                                                                                #
######################################################################################################################

import logging
from crhelper import CfnResource

from util.logging import get_logger
from util.quicksight import QuicksightApi

logger = logging.getLogger(__name__)
helper = CfnResource(json_logging=False, log_level='INFO')

def get_resource_propertices(event, _):
    logger.debug(f'servicing request event:{event}')
    request_type = event['RequestType']
    resource_properties = event['ResourceProperties']
    resource = resource_properties['Resource']

    logger.info(f'servicing request request_type:{request_type} resource:{resource}')
    # TODO: Implement more flexibility in loading config and/or state from first invocation and other invocations
    resource_properties.update({'ConfigDataFile': 'util/config/config.yaml'})
    return resource_properties

def log_exception(error):
    """
    Do logging in addition to crhelper exception handeling
    """
    logger.error(f'quicksight-application resources action error: {error}')
    import sys
    import traceback
    exc_type, exc_value, exc_traceback = sys.exc_info()
    format_exception = traceback.format_exception(
        exc_type, exc_value,
        exc_traceback
    )
    logger.error(repr(format_exception))


@ helper.create
def custom_resource_create(event, _):
    request_type = 'Create'
    resource_properties = get_resource_propertices(event, _)
    resource = resource_properties['Resource']
    qs_api = QuicksightApi(resource_properties)

    try:
        if resource == 'all':
            qs_api.create_all_resources()
        elif resource == 'datasource':
            qs_api.create_data_source()
        elif resource == 'dataset':
            qs_api.create_data_sets()
        elif resource == 'analysis':
            qs_api.create_analysis()
        elif resource == 'dashboard':
            qs_api.create_dashboard()
        else:
            logger.error(f'Not handling request resource:{resource}, request_type:{request_type}')
            raise ValueError(f'Received unsupported request request_type:{request_type}, resource:{resource}')
        if resource in ['all', 'analysis', 'dashboard']:
            analysis_url = qs_api.quicksight_application.get_analysis().url
            dashboard_url = qs_api.quicksight_application.get_dashboard().url
            helper.Data.update({
                'analysis_url': analysis_url,
                'dashboard_url': dashboard_url
            })
    except Exception as error:
        # Do logging in addition to crhelper exception handeling
        log_exception(error)
        qs_api.delete_all_resources()
        raise(error)

    logger.info(f'finished with request_type:{request_type} resource:{resource}')
    return None

@ helper.delete
def custom_resource_delete(event, _):
    request_type = 'Delete'
    resource_properties = get_resource_propertices(event, _)
    resource = resource_properties['Resource']
    qs_api = QuicksightApi(resource_properties)

    try:
        if resource == 'all':
            qs_api.delete_all_resources()
        elif resource == 'datasource':
            qs_api.delete_data_source()
        elif resource == 'dataset':
            qs_api.delete_data_sets()
        elif resource == 'analysis':
            qs_api.delete_analysis()
        elif resource == 'dashboard':
            qs_api.delete_dashboard()
        else:
            logger.error(f'Not handling request resource:{resource}, request_type:{request_type}')
            raise ValueError(f'Received unsupported request request_type:{request_type}, resource:{resource}')
    except Exception as error:
        # Do logging in addition to crhelper exception handeling
        log_exception(error)
        raise(error)

    logger.info(f'finished with request_type:{request_type} resource:{resource}')
    return None

@ helper.update
def custom_resource_update(event, _):
    # TODO: We don't support update. There are some updates that we should conceivably support. For example
    #       an update to the principal arn used when creating the resources
    try:
        request_type = 'Update'
        resource_properties = get_resource_propertices(event, _)
        resource = resource_properties['Resource']

        logger.error(f'Not handling request resource:{resource}, request_type:{request_type}')
        raise ValueError(f'Received unsupported request request_type:{request_type}, resource:{resource}')

    except Exception as error:
        # Do logging in addition to crhelper exception handeling
        log_exception(error)
        raise(error)

def handler(event, context):
    helper(event, context)
