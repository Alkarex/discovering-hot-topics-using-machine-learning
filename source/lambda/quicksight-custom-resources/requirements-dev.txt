moto==1.3.16
pytest-cov
pytest-env
pytest
pytest-mock
