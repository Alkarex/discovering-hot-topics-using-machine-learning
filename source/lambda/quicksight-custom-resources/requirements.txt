crhelper==2.0.6
requests==2.25.1
pyyaml
tenacity==6.2.0