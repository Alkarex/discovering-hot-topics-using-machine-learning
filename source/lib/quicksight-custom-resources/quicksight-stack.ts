#!/usr/bin/env node
/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import { NestedStack, CfnParameter, Construct, NestedStackProps } from '@aws-cdk/core';
import { QuickSight, QuickSightSetup } from './quicksight-construct';
import { ExecutionRole } from '../solution-helper/lambda-role-cloudwatch-construct';
export class QuickSightStack extends NestedStack {
    private _quickSight: QuickSight;

    constructor(scope: Construct, id: string, props?:NestedStackProps) {
        super(scope, id, props);

        /**
         * The QuickSight source template arn should follow the format below. The sourceTemplateAccountId is the aws
         * account id that is sharing the template and should have
         * `arn:${cdk.Aws.PARTITION}:quicksight:${cdk.Aws.REGION}:${sourceTemplateAccountId}:template/${sourceTemplateName}`
         * either granted public access to its template or specific access for this account to use it.
         */
        const quickSightSourceTemplateArn = new CfnParameter(this, 'QuickSightSourceTemplateArn', {
            type: 'String',
            description: 'The Amazon QuickSight template arn to use as the source template when creating analysis and dashboard resources',
            allowedPattern: '^arn:\\S+:quicksight:\\S+:\\d{12}:template/\\S+$',
            constraintDescription: 'Provide an arn matching an Amazon Quicksight template arn. The input did not match the validation pattern.'
        });

        /**
         * The QuickSight principal is used as the principal in the permissions of QuickSight resources created by the custom resources.
         * `arn:${cdk.Aws.PARTITION}:quicksight:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:user/${namespace}/${user_name`
         *
         * Example for namescape is 'default'. You can use aws cli command quicksight list-users and list-namespaces to choose from
         * Sample CLI command:
         * aws quicksight list-namespaces --aws-account-id $AWS_ACCOUNT_ID
         * aws quicksight list-users --aws-account-id $AWS_ACCOUNT_ID  --namespace default
         *
         */
        const quicksightPrincipalArn = new CfnParameter(this, 'QuickSightPrincipalArn', {
            type: 'String',
            description: 'The Amazon QuickSight principal arn used in the permissions of QuickSight resources',
            allowedPattern: '^arn:\\S+:quicksight:\\S+:\\d{12}:user/\\S+$',
            constraintDescription: 'Provide an arn matching an Amazon Quicksight principal arn. The input did not match the validation pattern.'
        });

        const s3AccessLogging = new CfnParameter(this, 'S3AccessLogBucket', {
            type: 'String',
            description: 'The S3 Access Log bucket arn',
            allowedPattern: 'arn:\\S+:s3:::\\S+$',
            constraintDescription: 'Please provide a valid S3 bucket arn to store S3 access logs'
        });

        const customResourceRole = new ExecutionRole(this, 'CustomResourceRole');

        this._quickSight = new QuickSight(this, 'Quicksight', {
            name: 'SO0122-Discovering-Hot-Topics-QS-Dashboard',
            resource: QuickSightSetup.ALL,
            sourceTemplateArn: quickSightSourceTemplateArn.valueAsString,
            principalArn: quicksightPrincipalArn.valueAsString,
            workgroupName: 'primary',
            logLevel: 'INFO',
            role: customResourceRole.Role
        });
    }

    public get analysisURLOutput(): string {
        return this._quickSight.analysisURL;
    }

    public get dashboardURLOutput(): string {
        return this._quickSight.dashboardURL;
    }
}
