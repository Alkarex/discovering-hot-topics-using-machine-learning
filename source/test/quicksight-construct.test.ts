/**********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICNSE-2.0                                                                     *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { QuickSight, QuickSightSetup } from '../lib/quicksight-custom-resources/quicksight-construct';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';

test('QS custom resource creation', () => {
    const stack = new Stack();
    new QuickSight(stack, 'testQSConstruct', {
        name: 'solution-name',
        resource: QuickSightSetup.ALL,
        sourceTemplateArn: 'arn:some-parition:quicksight:some-region:fakeaccountid:template/solution_solution-name-v_1_0',
        principalArn: 'arn:some-partition:quicksight:some-region:fakeaccountid:user/namespace/some-role/some-user',
        logLevel: 'INFO',
        workgroupName: 'testGroup',
        role: new Role(stack, 'testRole', {
            assumedBy: new ServicePrincipal('lambda.amazonaws.com')
        })
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    expect(stack).toHaveResourceLike('Custom::QuickSightResources', {
        "Resource": "all",
        "LogLevel": "INFO",
        "ApplicationName": "solution-name",
        "WorkGroupName": "testGroup"
    });

    expect(stack).toHaveResourceLike('AWS::IAM::Role', {
        "AssumeRolePolicyDocument": {
            "Statement": [{
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": {
                    "Service": "lambda.amazonaws.com"
                }
            }]
        }
    });

    expect(stack).toHaveResourceLike('AWS::Lambda::Function', {
        "Handler": "lambda_function.handler",
        "Runtime": "python3.8",
        "Timeout": 30
    });
});