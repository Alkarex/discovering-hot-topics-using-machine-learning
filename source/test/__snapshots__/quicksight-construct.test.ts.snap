// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QS custom resource creation 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5ccArtifactHash530DA107": Object {
      "Description": "Artifact hash for asset \\"47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5cc\\"",
      "Type": "String",
    },
    "AssetParameters47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5ccS3BucketF9DBAF3D": Object {
      "Description": "S3 bucket for asset \\"47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5cc\\"",
      "Type": "String",
    },
    "AssetParameters47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5ccS3VersionKeyA9CD1D5B": Object {
      "Description": "S3 key for asset version \\"47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5cc\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "testQSConstructCustomResource61E8811D": Object {
      "DependsOn": Array [
        "testQSConstructQSCustomResourcePolicy4A5596ED",
        "testRole836465CB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5ccS3BucketF9DBAF3D",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5ccS3VersionKeyA9CD1D5B",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters47bcd4ba11b0378fd41c5acc8c4730fb7fd860821abf9b9c19ce4f4ad263d5ccS3VersionKeyA9CD1D5B",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lambda_function.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "testRole836465CB",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "testQSConstructQSCustomResourcePolicy4A5596ED": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "The DescribeTemplate API call requires the resource to '*' in us-east-1.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "quicksight:CreateAnalysis",
                "quicksight:DeleteAnalysis",
                "quicksight:CreateDataSet",
                "quicksight:DeleteDataSet",
                "quicksight:CreateDataSource",
                "quicksight:DeleteDataSource",
                "quicksight:Describe*",
                "quicksight:Get*",
                "quicksight:List*",
                "quicksight:PassDataSet",
                "quicksight:PassDataSource",
                "quicksight:RestoreAnalysis",
                "quicksight:SearchAnalyses",
                "quicksight:CreateDashboard",
                "quicksight:DeleteDashboard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":quicksight:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":*/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "quicksight:DescribeTemplate",
              "Effect": "Allow",
              "Resource": "arn:some-parition:quicksight:some-region:fakeaccountid:template/solution_solution-name-v_1_0",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testQSConstructQSCustomResourcePolicy4A5596ED",
        "Roles": Array [
          Object {
            "Ref": "testRole836465CB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testQSConstructQuickSightResourcesA23CC3B0": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "testQSConstructQSCustomResourcePolicy4A5596ED",
      ],
      "Properties": Object {
        "ApplicationName": "solution-name",
        "LogLevel": "INFO",
        "QuickSightPrincipalArn": "arn:some-partition:quicksight:some-region:fakeaccountid:user/namespace/some-role/some-user",
        "QuickSightSourceTemplateArn": "arn:some-parition:quicksight:some-region:fakeaccountid:template/solution_solution-name-v_1_0",
        "Resource": "all",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "testQSConstructCustomResource61E8811D",
            "Arn",
          ],
        },
        "StackName": Object {
          "Ref": "AWS::StackName",
        },
        "WorkGroupName": "testGroup",
      },
      "Type": "Custom::QuickSightResources",
      "UpdateReplacePolicy": "Delete",
    },
    "testRole836465CB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
